File: teste_ml.rb
class Decompose
  def numbers(max_digit)
    valid_values = []
    max_value = max_digit + max_digit * 10 + max_digit * 100 + max_digit * 1000
    min_value = 1000

    x = min_value

    while x < max_value
      arr = x.to_s.split('')

      if arr[0].to_i <= max_digit && arr[1].to_i <= max_digit && arr[2].to_i <= max_digit && arr[3].to_i <= max_digit
        sum = arr[0].to_i + arr[1].to_i + arr[2].to_i + arr[3].to_i

        valid_values.push(arr.join('')) if sum == 21
      end
      x += 1
    end
    if valid_values.empty?
      puts 'Sem resultado'
    else
      puts valid_values
    end
  end
end

Decompose.new.call(8)

File: testeML.js
function decomposeNumbers(maxDigit) {
    let arrayValidValues = [];
    let maxValue = maxDigit + maxDigit * 10 + maxDigit * 100 + maxDigit * 1000;
    let minValue = 1000;

    for (let i = minValue; i < maxValue; i++) {
        let stringValue = i.toString();
        let arr = stringValue.split("");

        if ((arr[0] <= maxDigit) && (arr[1] <= maxDigit) && (arr[2] <= maxDigit) && arr[3] <= maxDigit) {

            let sum = parseInt(arr[0]) + parseInt(arr[1]) + parseInt(arr[2]) + parseInt(arr[3]);

            if (sum == 21) {
                arrayValidValues.push(arr.join(""));
            }

        }
    }
    if (arrayValidValues.length > 0) {
        for (i = 0; i < arrayValidValues.length; i++) {
            console.log(arrayValidValues[i]);
        }
    } else {
        console.log(null);
    }
}

decomposeNumbers(8);
